src/Middleware/OutputCaching/src/FormatterBinaryReader.cs

```diff
@@ -0,0 +1,197 @@
+
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.AspNetCore.OutputCaching;
+
+internal ref struct FormatterBinaryReader
+{
+    // this is effectively a cut-down re-implementation of BinaryReader
+    // from https://github.com/dotnet/runtime/blob/3689fbec921418e496962dc0ee252bdc9eafa3de/src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
+    // and is byte-compatible; however, instead of working against a Stream, we work against a ReadOnlyMemory<byte>
+    //
+    // additionally, we add support for reading a string with length specified by the caller (rather than handled automatically),
+    // and in-place (zero-copy) BLOB reads
+
+    private readonly ReadOnlyMemory<byte> original; // used to allow us to zero-copy chunks out of the payload
+    private readonly ref byte root;
+    private readonly int length;
+    private int offset;
+
+    public bool IsEOF => offset >= length;
+
+    public FormatterBinaryReader(ReadOnlyMemory<byte> content)
+    {
+        original = content;
+        length = content.Length;
+        root = ref MemoryMarshal.GetReference(content.Span);
+    }
+
+    public byte ReadByte()
+    {
+        if (IsEOF)
+        {
+            ThrowEndOfStream();
+        }
+        return Unsafe.Add(ref root, offset++);
+    }
+
+    public int Read7BitEncodedInt()
+    {
+        // Unlike writing, we can't delegate to the 64-bit read on
+        // 64-bit platforms. The reason for this is that we want to
+        // stop consuming bytes if we encounter an integer overflow.
+
+        uint result = 0;
+        byte byteReadJustNow;
+
+        // Read the integer 7 bits at a time. The high bit
+        // of the byte when on means to continue reading more bytes.
+        //
+        // There are two failure cases: we've read more than 5 bytes,
+        // or the fifth byte is about to cause integer overflow.
+        // This means that we can read the first 4 bytes without
+        // worrying about integer overflow.
+
+        const int MaxBytesWithoutOverflow = 4;
+        for (int shift = 0; shift < MaxBytesWithoutOverflow * 7; shift += 7)
+        {
+            // ReadByte handles end of stream cases for us.
+            byteReadJustNow = ReadByte();
+            result |= (byteReadJustNow & 0x7Fu) << shift;
+
+            if (byteReadJustNow <= 0x7Fu)
+            {
+                return (int)result; // early exit
+            }
+        }
+
+        // Read the 5th byte. Since we already read 28 bits,
+        // the value of this byte must fit within 4 bits (32 - 28),
+        // and it must not have the high bit set.
+
+        byteReadJustNow = ReadByte();
+        if (byteReadJustNow > 0b_1111u)
+        {
+            ThrowOverflowException();
+        }
+
+        result |= (uint)byteReadJustNow << (MaxBytesWithoutOverflow * 7);
+        return (int)result;
+    }
+
```

Comment:
If we assume that smaller magnitude numbers are more common than larger magnitude numbers, then zig-zag encoding is more efficient for signed integers. Casting works but it results in 5-byte encodings for the likely common `-1` case and 10-byte encodings for a -1 64-bit int. https://github.com/dotnet/orleans/blob/964ed094a5f77b3181c330a0cf33c58e19b48b85/src/Orleans.Serialization/Utilities/VarIntWriterExtensions.cs#L50

Reply:
in theory yes they are (it is a general purpose API), but: the heart of this needs to support pre-existing data written via `BinaryWriter`, so: we can't change the layout at least for the old usage; we could say "v2 always uses zigzag"; however, if we look at the callers it turns out that everything *is* positive, except perhaps for offset minutes which is more mercurial; in reality, I don't think we need to worry about this nuance; we *could* add unsigned methods, but... honestly, I think it is fine as it is

Reply:
It's not optimal but it's ok since this is not a general-purpose serializer